README.md
# Eventually Consistent Datatypes

This is an implementation of highly-scalable [eventually consistent
datatypes](https://dl.acm.org/citation.cfm?id=2911158) for Clojure.
When you do massively parallel updates on STM, the abort rate can
become prohibitive and you don't want to wait until your transactions
succeed. One example is the counter from the paper, where each thread
increments the counter in some inner loop. And you asynchronously
merge the thread-local var from time to time with the global state.


While I haven't had a use case for them yet, I found the research
interesting and wanted to benchmark the approach against the Clojure
STM. Consider the implementation experimental for now. I think the
datatypes become interesting in many-core scenarios where you have
more than 10 (maybe 100+) threads hammering on a datastructure. Since
massive parallelism is a primary design goal of Clojure, I think they
are worth investigating then.

## Bag

I am not sure about the access semantics for the bag, it would be very
well possible to use something different than sequences there,
e.g. sets or vectors, since I haven't used Cons cells (Clojure doesn't
use them much), but Clojure's persistent lists which also have
constant time concatenation due to lazyness.

You will still have `O(#elements)` worse case access time for other
collections though, but subsequent accesses could benefit from a
different datatype. I haven't implemented the OR-set from the paper
yet, as it is more complicated and I am not sure whether a binary tree
would perform well. Feel free to add it though :) or open an issue, if
you have a use-case.

## Usage

[![Clojars Project](http://clojars.org/es.topiq/ecdts/latest-version.svg)](http://clojars.org/es.topiq/ecdts)

The general idea is that you create a thread-local version of a
datatype for each of you threads and a global version with which you
merge in the desired intervals.

Example benchmark for the counter. You always have a global version,
similar to a Clojure atom, which you merge into from time to time. The
datatypes are dereferable to retrieve the current thread-local
value. Read the paper for details.

~~~clojure
(require '[ecdts.core :refer :all])

(def g (counter))

;; example benchmark for the counter
  (let [st (.getTime (java.util.Date.))
        ta (Thread. (fn []
                      (let [a (counter)]
                        (loop [i 0
                               j 0]
                          (when (< i 20000000)
                            (inc! a)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread A"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        tb (Thread. (fn []
                      (let [a (counter)]
                        (loop [i 0
                               j 0]
                          (when (< i 20000000)
                            (inc! a)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread B"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        tc (Thread. (fn []
                      (let [a (counter)]
                        (loop [i 0
                               j 0]
                          (when (< i 20000000)
                            (inc! a)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread C"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        td (Thread. (fn []
                      (let [a (counter)]
                        (loop [i 0
                               j 0]
                          (when (< i 20000000)
                            (inc! a)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread D"
                                 (- (.getTime (java.util.Date.))
                                    st)))))]
    (.start ta)
    (.start tb)
    (.start tc)
    (.start td))
~~~

It is roughly 3 times as fast as the STM on my machine.

The bag works similarly.

~~~clojure
(def g (add-only-bag))

  (let [st (.getTime (java.util.Date.))
        ta (Thread. (fn []
                      (let [a (add-only-bag)]
                        (loop [i 0
                               j 0]
                          (when (< i 2000000)
                            (conj! a i)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread A"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        tb (Thread. (fn []
                      (let [a (add-only-bag)]
                        (loop [i 0
                               j 0]
                          (when (< i 2000000)
                            (conj! a i)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread B"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        tc (Thread. (fn []
                      (let [a (add-only-bag)]
                        (loop [i 0
                               j 0]
                          (when (< i 2000000)
                            (conj! a i)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread C"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        td (Thread. (fn []
                      (let [a (add-only-bag)]
                        (loop [i 0
                               j 0]
                          (when (< i 2000000)
                            (conj! a i)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread D"
                                 (- (.getTime (java.util.Date.))
                                    st)))))]
    (.start ta)
    (.start tb)
    (.start tc)
    (.start td))
~~~

It is roughly 4 times faster on my machine.

## License

Copyright © 2016 Christian Weilbach

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-04-21
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-04-21
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/ecdts/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/ecdts/compare/0.1.0...0.1.1


test/ecdts/core_test.clj
(ns ecdts.core-test
  (:require [clojure.test :refer :all]
            [ecdts.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/ecdts/core.clj
(ns ecdts.core
  (:refer-clojure :exclude [conj!]))

(set! *warn-on-reflection* true)

(defprotocol PIncrementable
  "Implementation detail. Needed because we can only mutate (set!)
  inside of deftype for thread safety reasons."
  (-inc [this]))

(defprotocol PMergeable
  "Protocol to merge a thread-local version of a datatype into the global version"
  (-merge [global local]))

(deftype ECCounter [^:volatile-mutable ^long a ;; keep the global value synchronized
                    ^:unsynchronized-mutable ^long b]
  clojure.lang.IDeref
  (deref [this] (+ a b))
  PIncrementable
  (-inc [this] (set! b (inc b)))
  PMergeable
  (-merge [this other]
    (locking this
      (locking other
        (let [other ^ECCounter other
              g (+ a (.-b other))]
          (set! a g)
          (set! (.-b other) 0)
          (set! (.-a other) g)
          g)))))

(defn counter
  "Creates an eventual consistent increment only counter. Optionally
  you can initialize a thread-local version with a value. Derefencing
  and incrementing are not synchronized for performance reasons, so
  these operations are only valid in the thread-local context."
  ([] (counter 0))
  ([init-value]
   (ECCounter. 0 init-value)))

(defn inc!
  "Increment the counter. (in-place)"
  [^ECCounter counter]
  (-inc counter))

(defn merge!
  "Merge local state into global state and update local state to
  reflect global changes. (in-place)"
  [global local]
  (-merge global local))


(comment
  (def foo (ECCounter. 0 0))

  (def bar (ECCounter. 0 0))

  (-inc foo)
  (-inc bar)

  (-merge foo foo)
  (-merge foo bar)

  (-inc bar)

  (def g (counter))

  (let [st (.getTime (java.util.Date.))
        ta (Thread. (fn []
                      (let [a (counter)]
                        (loop [i 0
                               j 0]
                          (when (< i 20000000)
                            (inc! a)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread A"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        tb (Thread. (fn []
                      (let [a (counter)]
                        (loop [i 0
                               j 0]
                          (when (< i 20000000)
                            (inc! a)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread B"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        tc (Thread. (fn []
                      (let [a (counter)]
                        (loop [i 0
                               j 0]
                          (when (< i 20000000)
                            (inc! a)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread C"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        td (Thread. (fn []
                      (let [a (counter)]
                        (loop [i 0
                               j 0]
                          (when (< i 20000000)
                            (inc! a)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread D"
                                 (- (.getTime (java.util.Date.))
                                    st)))))]
    (.start ta)
    (.start tb)
    (.start tc)
    (.start td))

  @g

  (let [g (atom 0)
        st (.getTime (java.util.Date.))
        ta (Thread. (fn []
                      (loop [i 0]
                        (when (< i 20000000)
                          (swap! g inc)
                          (recur (inc i))))
                      (println "Thread A"
                               (- (.getTime (java.util.Date.))
                                  st))))
        tb (Thread. (fn []
                      (loop [i 0]
                        (when (< i 20000000)
                          (swap! g inc)
                          (recur (inc i))))
                      (println "Thread B"
                               (- (.getTime (java.util.Date.))
                                  st))))
        tc (Thread. (fn []
                      (loop [i 0]
                        (when (< i 20000000)
                          (swap! g inc)
                          (recur (inc i))))
                      (println "Thread C"
                               (- (.getTime (java.util.Date.))
                                  st))))
        td (Thread. (fn []
                      (loop [i 0]
                        (when (< i 20000000)
                          (swap! g inc)
                          (recur (inc i))))
                      (println "Thread D"
                               (- (.getTime (java.util.Date.))
                                  st))))]
    (.start ta)
    (.start tb)
    (.start tc)
    (.start td)))


(defprotocol PConjable
  "Implementation detail. Needed because we can only mutate (set!)
  inside of deftype for thread safety reasons."
  (-conj [this elem]))

(defn conj!
  "Adds element to an AddOnlyBag."
  [bag elem]
  (-conj bag elem))

(deftype ECAddOnlyBag [^:volatile-mutable l
                       ^:unsynchronized-mutable appender]
  clojure.lang.IDeref
  (deref [this] (concat appender l))
  PMergeable
  (-merge [this other]
    (locking this
      (locking other
        (let [other ^ECAddOnlyBag other
              new-list (concat (.-appender other) l)]
          (set! l new-list)
          (set! (.-l other) new-list)
          (set! (.-appender other) '())
          new-list))))
  PConjable
  (-conj [this elem] (set! appender (conj appender elem))))

(defn add-only-bag
  "Creates an AddOnlyBag. Optionally you can initialize a thread-local
  version with a seq value. Derefencing and conjing are not
  synchronized for performance reasons, so these operations are only
  valid in the thread-local context."
  ([] (add-only-bag '()))
  ([init-values]
   (ECAddOnlyBag. '() init-values)))

(comment
  (def foo (ECAddOnlyBag. '() '(1 2 3)))

  (def bar (ECAddOnlyBag. '() '()))

  (-conj foo 4)

  (-merge bar foo)

  (def g (add-only-bag))

  (let [st (.getTime (java.util.Date.))
        ta (Thread. (fn []
                      (let [a (add-only-bag)]
                        (loop [i 0
                               j 0]
                          (when (< i 2000000)
                            (conj! a i)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread A"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        tb (Thread. (fn []
                      (let [a (add-only-bag)]
                        (loop [i 0
                               j 0]
                          (when (< i 2000000)
                            (conj! a i)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread B"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        tc (Thread. (fn []
                      (let [a (add-only-bag)]
                        (loop [i 0
                               j 0]
                          (when (< i 2000000)
                            (conj! a i)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread C"
                                 (- (.getTime (java.util.Date.))
                                    st)))))
        td (Thread. (fn []
                      (let [a (add-only-bag)]
                        (loop [i 0
                               j 0]
                          (when (< i 2000000)
                            (conj! a i)
                            (if (> j 1000)
                              (do
                                (merge! g a)
                                (recur (inc i) 1))
                              (recur (inc i) (inc j)))))
                        (merge! g a)
                        (println "Thread D"
                                 (- (.getTime (java.util.Date.))
                                    st)))))]
    (.start ta)
    (.start tb)
    (.start tc)
    (.start td))

  (count @g)

  (let [g (atom '())
        st (.getTime (java.util.Date.))
        ta (Thread. (fn []
                      (loop [i 0]
                        (when (< i 2000000)
                          (swap! g conj i)
                          (recur (inc i))))
                      (println "Thread A"
                               (- (.getTime (java.util.Date.))
                                  st))))
        tb (Thread. (fn []
                      (loop [i 0]
                        (when (< i 2000000)
                          (swap! g conj i)
                          (recur (inc i))))
                      (println "Thread B"
                               (- (.getTime (java.util.Date.))
                                  st))))
        tc (Thread. (fn []
                      (loop [i 0]
                        (when (< i 2000000)
                          (swap! g conj i)
                          (recur (inc i))))
                      (println "Thread C"
                               (- (.getTime (java.util.Date.))
                                  st))))
        td (Thread. (fn []
                      (loop [i 0]
                        (when (< i 2000000)
                          (swap! g conj i)
                          (recur (inc i))))
                      (println "Thread D"
                               (- (.getTime (java.util.Date.))
                                  st))))]
    (.start ta)
    (.start tb)
    (.start tc)
    (.start td)))
